node_modules/col-assets/lib/api.js

```diff
@@ -187,3 +201,175 @@ var createLink = module.exports.createLink = function(ctx, title, url, opts, cal
+ * @param  {Object}         callback.err                    An error that occurred, if any
+ * @param  {Comment}        callback.comment                The created comment
+ * @param  {User}           callback.comment.user           The user that creted the comment
+ * @api private
+ */
+var getComment = function(id, callback) {
+  var options = {
+    'where': {
+      'id': id
+    },
+    'include': [{
+      'model': DB.User,
+      'attributes': ['id', 'canvas_user_id', 'canvas_course_role', 'canvas_full_name', 'canvas_image']
+    }]
+  };
+  DB.Comment.find(options).complete(function(err, comment) {
+    if (err) {
+      log.error({'err': err}, 'Failed to retrieve the created comment');
+      return callback({'code': 500, 'msg': err.message});
+    }
+
+    return callback(null, comment);
+  });
+};
+
+/**
+ * Create a new comment on an asset
+ *
+ * @param  {Context}        ctx                             Standard context containing the current user and the current course
+ * @param  {Number}         assetId                         The id of the asset to which the comment is added
+ * @param  {String}         body                            The body of the comment
+ * @param  {Number}         [parent]                        The id of the asset to which the comment is a reply
+ * @param  {Function}       callback                        Standard callback function
+ * @param  {Object}         callback.err                    An error that occurred, if any
+ * @param  {Comment}        callback.comment                The created comment
+ * @param  {User}           callback.comment.user           The user that created the comment
+ */
+var createComment = module.exports.createComment = function(ctx, assetId, body, parent, callback) {
+  // Parameter validation
+  var validationSchema = Joi.object().keys({
+    'assetId': Joi.number().required(),
+    'body': Joi.string().required(),
+    'parent': Joi.number().optional()
+  });
+
+  var validationResult = Joi.validate({
+    'assetId': assetId,
+    'body': body,
+    'parent': parent
+  }, validationSchema);
+
+  if (validationResult.error) {
+    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
+  }
+
+  // Get the asset to which the comment is being added
+  getAsset(ctx, assetId, false, function(err, asset) {
+    if (err) {
+      return callback(err);
+    }
+
+    // Check that the comment to which this comment is a reply exists
+    if (parent) {
+      // Ensure the parent id is a number
+      parent = CollabosphereUtil.getNumberParam(parent);
+      var parentComment = _.findWhere(asset.toJSON().comments, {'id': parent});
+      if (!parentComment) {
+        log.error({'err': err, 'parent': parent}, 'Failed to find the comment to which a reply is being made');
+        return callback({'code': 400, 'msg': 'Failed to find the comment to which a reply is being made'});
+      }
```

Comment:
Should we do a check to see if the `parentComment` has no `parent` of its own? IIRC we wouldn't allow for further nesting right?